### ISSUE TITLE
warn about unnecessary suppressions

### ISSUE THREAD

**Issue by cushon** [typetools/checker-framework#1701]
When upgrading a large code base to use the latest version of the checker it's sometimes necessary to add `@SuppressWarnings` for newly reported errors and investigate them later, instead of coordinating a large number of fixes with the upgrade. This makes it easy to accumulate suppressions, some of which were work-arounds for false positives that are unnecessary in the latest version of the checker. Leaving the annotations in place clutters up code and makes the checker less effective.

It would be helpful to have a mode where the checker could report unnecessary `@SuppressWarnings` annotations.

**Comment by mernst** [comment]
This would be a nice feature.
As a workaround, I do it this way:  Whenever I suppress a warning, I always write a comment on the same line with a link to the relevant issue.  For example: [here](https://github.com/mernst/plume-lib/blob/master/java/src/plume/ArraysMDE.java#L2755) and [here](https://github.com/mernst/plume-lib/blob/master/java/src/plume/ArraysMDE.java#L3454).

Periodically, I visit the URL to see whether the issue has been closed and remove the suppressions that are no longer necessary.  (I don't have the program that does this, because it only takes a minute to write and I rewrite it each time...)

**Comment by mernst** [comment]
Another workaround is to remove all the warning suppressions and then reinstate the ones that are needed.  Reinstating an annotation is a keystroke in an IDE if you are using version control.

I realize that you are running on much larger codebases that make manual approaches unrealistic (you would need to do this programmatically), but I wanted to mention it for other people who are reading the issue.

**Comment by mernst** [comment]
Here is how the feature could be implemented:

 * Keep a stack of all the warning suppressions that are currently active.
 * Whenever a warning is issued, if it is stopped by a warning suppression, mark it.
 * When leaving the scope of a warning suppression, if it is not marked, output it as unnecessary.

At some point -- maybe when adding to the stack, maybe when doing output -- check that the warning suppression is relevant to the checkers being run.  We don't want to warn about `@SuppressWarnings` that are relevant to regular Java compilation, or Eclipse, or Error Prone.

(It might be enough to keep a list of all warning suppressions ever encountered, rather than the stack?  Remembering that whole list might be inefficient on large codebases, though.)

We should implement this when we get a chance.  It might be a good starter project for someone.

**Comment by cushon** [comment]
Thanks - I agree about the workarounds. In my case I'm sometimes seeing more issues than I have time to investigate individually, so I make guesses about root causes and look at a subset more closely. So it's not always clear later which issue a particular suppression is blocked on.

Removing all of the suppressions that correspond to checker framework warnings and then adding back the ones that are still needed programmatically is a good suggestions, it should be easy to make that scale.

Built-in support for detecting unnecessary suppressions would still be nice to have, though.

**Comment by smillst** [comment]
It is best practice if you are using more than one checker on the same code and/or if the warning key could apply to more than one checker.  (We are going to do a few things to make this clear in the July release.)

I'll change the tool to look at any `@SuppressWarnings` with a key that might have been issued by the checker.  (That is any key in its messages.properties file or any messages.properties that applies to the checker.)
